airbyte-featureflag/src/main/kotlin/Client.kt

```diff
@@ -0,0 +1,196 @@
+import com.fasterxml.jackson.module.kotlin.readValue
+import com.fasterxml.jackson.module.kotlin.registerKotlinModule
+import com.launchdarkly.sdk.ContextKind
+import com.launchdarkly.sdk.LDContext
+import com.launchdarkly.sdk.LDUser
+import com.launchdarkly.sdk.server.LDClient
+import java.lang.Thread.MIN_PRIORITY
+import java.nio.file.Path
+import java.nio.file.StandardWatchEventKinds
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.thread
+import kotlin.concurrent.write
+import kotlin.io.path.isRegularFile
+
+/**
+ * Open Questions:
+ * 1. Can the context and flag be combined into one class?
+ *   - i.e. Does it make sense to one a feature-flag apply to two different context types?
+ * 2. Should the default value be merged into the Flag type instead of specified on each check?
+ *   - Or maybe still allow it to be specified, but make it an optional parameter
+ */
+
+/**
+ * Feature-Flag Client interface.
+ */
+sealed interface FeatureFlagClient {
+    /**
+     * Returns true if the flag with the provided context should be enabled. Returns false otherwise.
+     */
+    fun enabled(flag: Flag, ctx: Context): Boolean
+}
+
+/**
+ * Platform feature-flag client. Feature-flag are derived from a yaml config file.
+ * Also supports flags defined via environment-variables via the [EnvVar] class.
+ *
+ * @param [config] the location of the yaml config file that contains the feature-flag definitions.
+ * The [config] will be watched for changes and the internal representation of the [config] will be updated to match.
+ */
+class PlatformClient(config: Path) : FeatureFlagClient {
+    /** [flags] holds the mappings of the flag-name to the flag properties */
+    private var flags: Map<String, PlatformFlag> = readConfig(config)
+
+    /** lock is used for ensuring access to the flags map is handled correctly when the map is being updated. */
+    private val lock = ReentrantReadWriteLock()
+
+    init {
+        if (!config.isRegularFile()) {
+            throw IllegalArgumentException("config must reference a file")
+        }
+
+        config.onChange {
+            val tempFlags = readConfig(config)
+            lock.write { flags = tempFlags }
+        }
+    }
+
+    override fun enabled(flag: Flag, ctx: Context): Boolean {
+        return when (flag) {
+            is EnvVar -> flag.enabled()
+            else -> lock.read { flags[flag.key]?.enabled ?: flag.default }
+        }
+    }
+}
+
+/**
+ * Cloud feature-flag client. Feature-flags are derived from an external source (the LDClient).
+ * Also supports flags defined via environment-variables via the [EnvVar] class.
+ *
+ * @param [client] the Launch-Darkly client for interfacing with Launch-Darkly.
+ */
+class CloudClient(private val client: LDClient) : FeatureFlagClient {
+    override fun enabled(flag: Flag, ctx: Context): Boolean {
+        return when (flag) {
+            is EnvVar -> flag.enabled()
```

Comment:
it's a little confusing to me why the caller of `enabled` knows if the flag is if an EnvFlag or not. i assumed the point of the abstraction was that the caller has a "flag" (roughly a string) and wants to know if it's turn on or not, but doesn't need to know where that flag is set. passing in `EnvFlag` breaks that abstraction. how are you thinking about it.

Reply:
> it's a little confusing to me why the caller of enabled knows if the flag is if an EnvFlag or not

The caller shouldn't care if the flag they're using is an `EnvVar` flag or not.  The idea is that all flags are defined as singletons in a central location and the caller would simply reference the specific flags by their singleton, ignoring what type they actually are.

E.g. we have the following flag currently defined
```kotlin
object LogConnectorMessages : EnvVar(envVar = "LOG_CONNECTOR_MESSAGES", team = Team.PLATFORM_MOVE)
```
which currently is an `EnvVar` type flag, but the user of this flag doesn't know that when providing it to the `enabled` method, they simply call
``` kotlin
val client: FeatureFlagClient = /* whatever implementation they're using */
client.enabled(LogConnectorMessages, WorkspaceContext("workspace-id")
```
As these flags are migrated away from environment-variables, the definition will be updated to 
```kotlin
object LogConnectorMessages : Flag(key = "LOG_CONNECTOR_MESSAGES", team = Team.PLATFORM_MOVE)
```
and the caller will be none the wiser.

The `FeatureFlagClient` implementations themselves need to be aware if the Flag is of type `EnvVar` as environment-variables based flags will never be found in the `flags.yml` file or in the LaunchDarkly API.  The client needs to know that these flags are special cases and require special handling.  This is why the `EnvVar.enabled()` method is defined as `internal`, it isn't visible to any code outside of this module.